### ISSUE TITLE
Jumping to strange address after call to LoadLibraryA

### ISSUE THREAD

**Issue by tomchop** [cea-sec/miasm#65]
I'm getting a strange behavior when calling `LoadLibraryA` (or more precisely from the `exec_wrapper` function)

```
cannot find crypto, skipping
add module 0x400000 'd\x00l\x00u\x00h\x00_\x002\x000\x001\x005\x00-\x000\x002\x00-\x000\x009\x00_\x001\x001\x00-\x003\x003\x00-\x003\x008\x00.\x00e\x00x\x00e\x00'
0x38 'd\x00l\x00u\x00h\x00_\x002\x000\x001\x005\x00-\x000\x002\x00-\x000\x009\x00_\x001\x001\x00-\x003\x003\x00-\x003\x008\x00.\x00e\x00x\x00e\x00'
add module 0x0 '\x00'
0x1 '\x00'
add module 0x7c910000 'n\x00t\x00d\x00l\x00l\x00.\x00d\x00l\x00l\x00'
0x12 'n\x00t\x00d\x00l\x00l\x00.\x00d\x00l\x00l\x00'
add module 0x7c800000 'k\x00e\x00r\x00n\x00e\x00l\x003\x002\x00.\x00d\x00l\x00l\x00'
0x18 'k\x00e\x00r\x00n\x00e\x00l\x003\x002\x00.\x00d\x00l\x00l\x00'
add module 0x7e390000 'u\x00s\x00e\x00r\x003\x002\x00.\x00d\x00l\x00l\x00'
0x14 'u\x00s\x00e\x00r\x003\x002\x00.\x00d\x00l\x00l\x00'
add module 0x774a0000 'o\x00l\x00e\x003\x002\x00.\x00d\x00l\x00l\x00'
0x12 'o\x00l\x00e\x003\x002\x00.\x00d\x00l\x00l\x00'
add module 0x7df30000 'u\x00r\x00l\x00m\x00o\x00n\x00.\x00d\x00l\x00l\x00'
0x14 'u\x00r\x00l\x00m\x00o\x00n\x00.\x00d\x00l\x00l\x00'
add module 0x71ab0000 'w\x00s\x002\x00_\x003\x002\x00.\x00d\x00l\x00l\x00'
0x14 'w\x00s\x002\x00_\x003\x002\x00.\x00d\x00l\x00l\x00'
add module 0x77f50000 'a\x00d\x00v\x00a\x00p\x00i\x003\x002\x00.\x00d\x00l\x00l\x00'
0x18 'a\x00d\x00v\x00a\x00p\x00i\x003\x002\x00.\x00d\x00l\x00l\x00'
add module 0x76bf0000 'p\x00s\x00a\x00p\x00i\x00.\x00d\x00l\x00l\x00'
0x12 'p\x00s\x00a\x00p\x00i\x00.\x00d\x00l\x00l\x00'
fix inloadorder
dluh_2015-02-09_11-33-38.exe

ntdll.dll
kernel32.dll
user32.dll
ole32.dll
urlmon.dll
ws2_32.dll
advapi32.dll
psapi.dll
dluh_2015-02-09_11-33-38.exe

ntdll.dll
kernel32.dll
user32.dll
ole32.dll
urlmon.dll
ws2_32.dll
advapi32.dll
psapi.dll
inloadorder first 0x341f00
ntdll 0x343f00
Start emulation 0x401e8b
[DEBUG]: my_GetEnvironmentVariable 0x401eee ['0x40727a', '0x407284', '0xa']
WARNING: WARNING unknown env variable 'SWkBlRevt'
[DEBUG]: kernel32_GetPrivateProfileStructW 0x401f2e ['0x407228', '0x407228', '0x407239']
[DEBUG]: kernel32_GetLongPathNameA 0x40146d ['0x4072ea', '0x4072fa', '0xd']
[DEBUG]: kernel32_GetPrivateProfileStructW 0x401dae ['0x407228', '0x407228', '0x407239']
[DEBUG]: kernel32_GetShortPathNameW 0x401dfe ['0x4072b2', '0x4072c2', '0x9']
[DEBUG]: kernel32_SetCurrentDirectoryW 0x401b91 ['0x407350']
[DEBUG]: kernel32_GetPrivateProfileStructW 0x40294d ['0x407228', '0x407228', '0x407239']
[DEBUG]: kernel32_GetPrivateProfileStructW 0x402989 ['0x407307', '0x407307', '0x407317']
[DEBUG]: kernel32_FindResourceA 0x4019a9 ['0x0', '0x40728f', '0x4072a1']
[DEBUG]: kernel32_SetCurrentDirectoryW 0x4019e2 ['0x407350']
[DEBUG]: kernel32_QueryDosDeviceW 0x401cc3 ['0x40725a', '0x40726a', '0x10']
[DEBUG]: kernel32_QueryDosDeviceW 0x401f9a ['0x40725a', '0x40726a', '0x10']
[DEBUG]: kernel32_GetPrivateProfileStructW 0x40117f ['0x407228', '0x407228', '0x407239']
[DEBUG]: kernel32_GetPrivateProfileStructW 0x402020 ['0x407228', '0x407228', '0x407239']
[DEBUG]: kernel32_WaitForSingleObject 0x4015bc ['0xffffffff', '0x3e8']
[DEBUG]: kernel32_LoadLibraryA 0x4016ac ['0x407210']
uxtheme.dll
INFO : uxtheme.dll
INFO : ret 71122000
[DEBUG]: kernel32_VirtualAlloc 0x402099 ['0x0', '0x674', '0x1000', '0x40']
[DEBUG]: kernel32_LoadLibraryA 0x2000000e ['0x200001bc']
Kernel32.dll
INFO : Kernel32.dll
INFO : ret 7c800000
[DEBUG]: kernel32_GetProcAddress 0x20000042 ['0x7c800000', '0x200001c9']
INFO : HeapAlloc
[DEBUG]: kernel32_GetProcAddress 0x2000004c ['0x7c800000', '0x200001d3']
INFO : HeapFree
[DEBUG]: kernel32_GetProcAddress 0x20000057 ['0x7c800000', '0x200001dc']
INFO : GetTickCount
[DEBUG]: kernel32_VirtualAlloc 0x2000006b ['0x0', '0x674', '0x1000', '0x40']
[DEBUG]: kernel32_VirtualAlloc 0x200015e2 ['0x0', '0x2000', '0x1000', '0x4']
[DEBUG]: kernel32_VirtualProtect 0x200010b9 ['0x400000', '0x200', '0x4', '0x132fef0']
[DEBUG]: kernel32_VirtualProtect 0x200010c9 ['0x400000', '0x200', '0x2', '0x132fef0']
[DEBUG]: kernel32_VirtualProtect 0x20001112 ['0x401000', '0x2000', '0x4', '0x132fed0']
[DEBUG]: kernel32_VirtualProtect 0x20001128 ['0x401000', '0x2000', '0x40', '0x132fed0']
[DEBUG]: kernel32_VirtualProtect 0x20001112 ['0x403000', '0x1000', '0x4', '0x132fed0']
[DEBUG]: kernel32_VirtualProtect 0x20001128 ['0x403000', '0x1000', '0x2', '0x132fed0']
[DEBUG]: kernel32_GetModuleHandle 0x2000120b ['0x40204c']
INFO : KERNEL32.dll
[DEBUG]: kernel32_VirtualProtect 0x20001230 ['0x401000', '0x54', '0x4', '0x132fec4']
[DEBUG]: kernel32_GetProcAddress 0x2000125b ['0x7c800000', '0x401f1a']
INFO : ExitProcess
[DEBUG]: kernel32_GetProcAddress 0x2000125b ['0x7c800000', '0x401f28']
INFO : FindResourceA
[DEBUG]: kernel32_GetProcAddress 0x2000125b ['0x7c800000', '0x401f38']
INFO : LoadResource
[DEBUG]: kernel32_GetProcAddress 0x2000125b ['0x7c800000', '0x401f48']
INFO : VirtualFree
[DEBUG]: kernel32_GetProcAddress 0x2000125b ['0x7c800000', '0x401f56']
INFO : WriteFile
[DEBUG]: kernel32_GetProcAddress 0x2000125b ['0x7c800000', '0x401f62']
INFO : Sleep
[DEBUG]: kernel32_GetProcAddress 0x2000125b ['0x7c800000', '0x401f6a']
INFO : SizeofResource
[DEBUG]: kernel32_GetProcAddress 0x2000125b ['0x7c800000', '0x401f7c']
INFO : GetModuleFileNameW
[DEBUG]: kernel32_GetProcAddress 0x2000125b ['0x7c800000', '0x401f92']
INFO : CreateFileW
[DEBUG]: kernel32_GetProcAddress 0x2000125b ['0x7c800000', '0x401fa0']
INFO : lstrlenW
[DEBUG]: kernel32_GetProcAddress 0x2000125b ['0x7c800000', '0x401fac']
INFO : GetTempPathW
[DEBUG]: kernel32_GetProcAddress 0x2000125b ['0x7c800000', '0x401fbc']
INFO : GetLastError
[DEBUG]: kernel32_GetProcAddress 0x2000125b ['0x7c800000', '0x401fcc']
INFO : GetProcAddress
[DEBUG]: kernel32_GetProcAddress 0x2000125b ['0x7c800000', '0x401fde']
INFO : VirtualAlloc
[DEBUG]: kernel32_GetProcAddress 0x2000125b ['0x7c800000', '0x401fee']
INFO : LoadLibraryA
[DEBUG]: kernel32_GetProcAddress 0x2000125b ['0x7c800000', '0x401ffe']
INFO : LockResource
[DEBUG]: kernel32_GetProcAddress 0x2000125b ['0x7c800000', '0x40200e']
INFO : GetModuleHandleA
[DEBUG]: kernel32_GetProcAddress 0x2000125b ['0x7c800000', '0x402022']
INFO : CreateMutexA
[DEBUG]: kernel32_GetProcAddress 0x2000125b ['0x7c800000', '0x402032']
INFO : CloseHandle
[DEBUG]: kernel32_GetProcAddress 0x2000125b ['0x7c800000', '0x402040']
INFO : DeleteFileW
[DEBUG]: kernel32_VirtualProtect 0x20001269 ['0x401000', '0x54', '0x40', '0x132fecc']
[DEBUG]: kernel32_GetModuleHandle 0x2000120b ['0x402068']
INFO : USER32.dll
[DEBUG]: kernel32_VirtualProtect 0x20001230 ['0x401054', '0x8', '0x4', '0x132fec4']
[DEBUG]: kernel32_GetProcAddress 0x2000125b ['0x7e390000', '0x40205c']
INFO : MessageBoxA
[DEBUG]: kernel32_VirtualProtect 0x20001269 ['0x401054', '0x8', '0x40', '0x132fecc']
[DEBUG]: kernel32_VirtualFree 0x2000118b ['0x20002000', '0x2000', '0x4000']
[DEBUG]: kernel32_LoadLibraryA 0x401956 ['0x4012a8']
SHLWAPI.DLL
INFO : SHLWAPI.DLL
INFO : ret 71119000
[DEBUG]: kernel32_LoadLibraryA 0x40195f ['0x4012b4']
SETUPAPI.DLL
INFO : SETUPAPI.DLL
INFO : ret 71123000
RAX 0000000071123000 RBX 0000000071119000 RCX 0000000000000000 RDX 0000000000000000
RSI 000000007C801D7B RDI 000000000132FFE1 RSP 000000000132F5A8 RBP 000000000133000C
zf 0000000000000000 nf 0000000000000000 of 0000000000000000 cf 0000000000000000
RIP 000000000040195F
0040195F PUSH       0x4012C4
RAX 0000000071123000 RBX 0000000071119000 RCX 0000000000000000 RDX 0000000000000000
RSI 000000007C801D7B RDI 000000000132FFE1 RSP 000000000132F5A4 RBP 000000000133000C
zf 0000000000000000 nf 0000000000000000 of 0000000000000000 cf 0000000000000000
RIP 000000000040195F
00401964 MOV        EBP, EAX
RAX 0000000071123000 RBX 0000000071119000 RCX 0000000000000000 RDX 0000000000000000
RSI 000000007C801D7B RDI 000000000132FFE1 RSP 000000000132F5A4 RBP 0000000071123000
zf 0000000000000000 nf 0000000000000000 of 0000000000000000 cf 0000000000000000
RIP 0000000000401964
00401966 CALL       ESI
[DEBUG]: kernel32_LoadLibraryA 0x401968 ['0x4012c4']
SHELL32.DLL
INFO : SHELL32.DLL
INFO : ret 71124000
--Return--
> /scripts/sandbox.py(102)kernel32_LoadLibraryA()->None
-> pdb.set_trace()
(Pdb) s   
> /usr/local/lib/python2.7/dist-packages/miasm2/jitter/jitload.py(369)handle_lib()
-> jitter.pc = getattr(jitter.cpu, jitter.ir_arch.pc.name)
(Pdb) s
> /usr/local/lib/python2.7/dist-packages/miasm2/jitter/jitload.py(370)handle_lib()
-> return True
(Pdb) s
--Return--
> /usr/local/lib/python2.7/dist-packages/miasm2/jitter/jitload.py(370)handle_lib()->True
-> return True
(Pdb) s
> /usr/local/lib/python2.7/dist-packages/miasm2/jitter/jitload.py(85)call_callbacks()
-> if res is not True:
(Pdb) s
> /usr/local/lib/python2.7/dist-packages/miasm2/jitter/jitload.py(83)call_callbacks()
-> for c in self.get_callbacks(name):
(Pdb) s
--Return--
> /usr/local/lib/python2.7/dist-packages/miasm2/jitter/jitload.py(83)call_callbacks()->None
-> for c in self.get_callbacks(name):
(Pdb) s
> /usr/local/lib/python2.7/dist-packages/miasm2/jitter/jitload.py(265)runiter_once()
-> if old_pc != self.pc:
(Pdb) s
> /usr/local/lib/python2.7/dist-packages/miasm2/jitter/jitload.py(266)runiter_once()
-> return
(Pdb) s
--Return--
> /usr/local/lib/python2.7/dist-packages/miasm2/jitter/jitload.py(266)runiter_once()->None
-> return
(Pdb) s
StopIteration: None
> /usr/local/lib/python2.7/dist-packages/miasm2/jitter/jitload.py(297)continue_run()
-> return self.run_iterator.next()
(Pdb) s
> /usr/local/lib/python2.7/dist-packages/miasm2/jitter/jitload.py(298)continue_run()
-> except StopIteration:
(Pdb) s
> /usr/local/lib/python2.7/dist-packages/miasm2/jitter/jitload.py(299)continue_run()
-> pass
(Pdb) s
> /usr/local/lib/python2.7/dist-packages/miasm2/jitter/jitload.py(301)continue_run()
-> self.run_iterator = self.runiter_once(self.pc)
(Pdb) s
> /usr/local/lib/python2.7/dist-packages/miasm2/jitter/jitload.py(303)continue_run()
-> if step is True:
(Pdb) s
> /usr/local/lib/python2.7/dist-packages/miasm2/jitter/jitload.py(295)continue_run()
-> while self.run:
(Pdb) s
> /usr/local/lib/python2.7/dist-packages/miasm2/jitter/jitload.py(296)continue_run()
-> try:
(Pdb) s
> /usr/local/lib/python2.7/dist-packages/miasm2/jitter/jitload.py(297)continue_run()
-> return self.run_iterator.next()
(Pdb) s
--Call--
> /usr/local/lib/python2.7/dist-packages/miasm2/jitter/jitload.py(246)runiter_once()
-> def runiter_once(self, pc):
(Pdb) s
> /usr/local/lib/python2.7/dist-packages/miasm2/jitter/jitload.py(250)runiter_once()
-> self.pc = pc
(Pdb) s
> /usr/local/lib/python2.7/dist-packages/miasm2/jitter/jitload.py(253)runiter_once()
-> if self.exec_cb is not None:
(Pdb) pc
4200808L
(Pdb) s
> /usr/local/lib/python2.7/dist-packages/miasm2/jitter/jitload.py(259)runiter_once()
-> old_pc = self.pc
(Pdb) s
> /usr/local/lib/python2.7/dist-packages/miasm2/jitter/jitload.py(260)runiter_once()
-> for res in self.breakpoints_handler(self.pc, self):
(Pdb) s
--Call--
> /usr/local/lib/python2.7/dist-packages/miasm2/jitter/jitload.py(88)__call__()
-> def __call__(self, name, *args):
(Pdb) s
> /usr/local/lib/python2.7/dist-packages/miasm2/jitter/jitload.py(90)__call__()
-> return self.call_callbacks(name, *args)
(Pdb) s
--Return--
> /usr/local/lib/python2.7/dist-packages/miasm2/jitter/jitload.py(90)__call__()-><generat...51e1a140>
-> return self.call_callbacks(name, *args)
(Pdb) s
--Call--
> /usr/local/lib/python2.7/dist-packages/miasm2/jitter/jitload.py(76)call_callbacks()
-> def call_callbacks(self, name, *args):
(Pdb) s
> /usr/local/lib/python2.7/dist-packages/miasm2/jitter/jitload.py(81)call_callbacks()
-> res = True
(Pdb) s
> /usr/local/lib/python2.7/dist-packages/miasm2/jitter/jitload.py(83)call_callbacks()
-> for c in self.get_callbacks(name):
(Pdb) s
--Call--
> /usr/local/lib/python2.7/dist-packages/miasm2/jitter/jitload.py(52)get_callbacks()
-> def get_callbacks(self, name):
(Pdb) s
> /usr/local/lib/python2.7/dist-packages/miasm2/jitter/jitload.py(54)get_callbacks()
-> return self.callbacks.get(name, [])
(Pdb) s
--Return--
> /usr/local/lib/python2.7/dist-packages/miasm2/jitter/jitload.py(54)get_callbacks()->[]
-> return self.callbacks.get(name, [])
(Pdb) s
--Return--
> /usr/local/lib/python2.7/dist-packages/miasm2/jitter/jitload.py(83)call_callbacks()->None
-> for c in self.get_callbacks(name):
(Pdb) s
> /usr/local/lib/python2.7/dist-packages/miasm2/jitter/jitload.py(265)runiter_once()
-> if old_pc != self.pc:
(Pdb) self.pc
4200808L
(Pdb) old_pc
4200808L
(Pdb) s
> /usr/local/lib/python2.7/dist-packages/miasm2/jitter/jitload.py(269)runiter_once()
-> assert(self.get_exception() == 0)
(Pdb) s
--Call--
> /usr/local/lib/python2.7/dist-packages/miasm2/jitter/jitload.py(317)get_exception()
-> def get_exception(self):
(Pdb) s
> /usr/local/lib/python2.7/dist-packages/miasm2/jitter/jitload.py(318)get_exception()
-> return self.cpu.get_exception() | self.vm.get_exception()
(Pdb) s
--Return--
> /usr/local/lib/python2.7/dist-packages/miasm2/jitter/jitload.py(318)get_exception()->0L
-> return self.cpu.get_exception() | self.vm.get_exception()
(Pdb) s
> /usr/local/lib/python2.7/dist-packages/miasm2/jitter/jitload.py(272)runiter_once()
-> self.pc = self.runbloc(self.pc)
(Pdb) s
--Call--
> /usr/local/lib/python2.7/dist-packages/miasm2/jitter/jitload.py(240)runbloc()
-> def runbloc(self, pc):
(Pdb) pc
4200808L
(Pdb) s
> /usr/local/lib/python2.7/dist-packages/miasm2/jitter/jitload.py(244)runbloc()
-> return self.jit.runbloc(self.cpu, self.vm, pc)
(Pdb) hex(pc)
'0x401968L' # this is the correct value
(Pdb) s
--Call--
> /usr/local/lib/python2.7/dist-packages/miasm2/jitter/jitcore.py(160)runbloc()
-> def runbloc(self, cpu, vm, lbl):
(Pdb) s
> /usr/local/lib/python2.7/dist-packages/miasm2/jitter/jitcore.py(167)runbloc()
-> if lbl is None:
(Pdb) s
> /usr/local/lib/python2.7/dist-packages/miasm2/jitter/jitcore.py(170)runbloc()
-> if not lbl in self.lbl2jitbloc:
(Pdb) s
> /usr/local/lib/python2.7/dist-packages/miasm2/jitter/jitcore.py(175)runbloc()
-> ret = self.jit_call(lbl, cpu, vm)
(Pdb) s
--Call--
> /usr/local/lib/python2.7/dist-packages/miasm2/jitter/jitcore.py(150)jit_call()
-> def jit_call(self, label, cpu, vmmngr):
(Pdb) s
> /usr/local/lib/python2.7/dist-packages/miasm2/jitter/jitcore.py(157)jit_call()
-> fc_ptr = self.lbl2jitbloc[label]
(Pdb) s
> /usr/local/lib/python2.7/dist-packages/miasm2/jitter/jitcore.py(158)jit_call()
-> return self.exec_wrapper(fc_ptr, cpu.cpu, vmmngr.vmmngr)
(Pdb) s
--Return-- # exec_wrapper returns 143017204 = 0x88644F4
> /usr/local/lib/python2.7/dist-packages/miasm2/jitter/jitcore.py(158)jit_call()->143017204L
-> return self.exec_wrapper(fc_ptr, cpu.cpu, vmmngr.vmmngr)
(Pdb) hex(143017204L)
'0x88644f4L'
(Pdb) c
WARNING: address 0x88644F4 is not mapped in virtual memory:
WARNING: [Errno cannot get mem ad] 0x88644f4L
WARNING: cannot disasm at 88644F4
> /scripts/sandbox.py(286)<module>()
-> assert(sb.jitter.run is False)
```
Here, `CALL ESI` is calling LoadLibraryA with parameter SHELL32.DLL. I overrid `kernel32_LoadLibraryA` with the following, to load some parts of `uxtheme.dll` which are verified by the binary in memory

```
def kernel32_LoadLibraryA(jitter):
    ret_ad, args = jitter.func_args_stdcall(1)
    dllname, = args

    libname = get_str_ansi(jitter, dllname, 0x100)
    print libname
    log.info(libname)

    ret = winobjs.runtime_dll.lib_get_add_base(libname)
    log.info("ret %x" %ret)

    # modifications here
    if libname.lower() == 'uxtheme.dll':
        with open('./win_dll/%s'%libname.lower()) as dllfile:
            dllcontents = dllfile.read()
            jitter.vm.add_memory_page(ret, PAGE_EXEC | PAGE_READ, dllcontents[:0x1000])

    jitter.func_ret_stdcall(ret_ad, ret)
```

Since this might be a little obscure to understand, I can share my script and sample if you need to.

**Comment by serpilliere** [comment]
Hi!

It seems you do not use the correct API. I cannot blame you, the documentation is not clear on this point. :wink:
The `winobjs.runtime_dll.lib_get_add_base(libname)` behaviour is to look for the `libname` in its database of loaded libraries. If an entry is found, it returns the base address of the found object. If it's not found, it will create a dummy entry, and add it to the database. In this case, it will return a dummy address.

The second problem is the way you load a dll in memory. You have to use the miasm PE loader instead of simply mapping bytes in memory. You should use `vm_load_pe_lib` in `miasm2/jitter/loader/pe.py`.

There is a simplest solution using the `-l` option of the script `sandbox_pe_x86_32.py`, which will load in memory dll from the directory "win_dll" as described in the help:
```
  -l, --loadbasedll     Load base dll (path './win_dll')
```

For instance, if you use this option it will load default dll:
```
class OS_Win(OS):
    # DLL to import
    ALL_IMP_DLL = ["ntdll.dll", "kernel32.dll", "user32.dll",
                   "ole32.dll", "urlmon.dll",
                   "ws2_32.dll", 'advapi32.dll', "psapi.dll",
               ]
```

But you can add your own dll to this list. In this case, the dll will be loaded before the execution, and the `runtime_dll` will be aware of its base address, and so will return its real address.

Last point, the default behaviour will be to try to call an "emulated" version of its exported function, so if your sample calls for example the function "toto", it will try to call `uxtheme_toto` in python.

**Comment by tomchop** [comment]
Hi, thanks for your quick answer!

I'm already using the `--loadbasedll` option, sorry for not saying so in the first place. But you're right, the DLL is not on the ALL_IMP_DLL.

FYI, the `kernel32_LoadLibraryA` is identical to the one that's already defined, except for the `# modifications here` block. I didn't care about correctly loading the PE (`uxtheme.dll`) in memory for now since the binary just checks a few bytes, but does not seem to call any functions from it (probably just a dummy check). The changes I brought only work for this particular DLL.

What I find strange is that other libraries such as `SHLWAPI.DLL` and `SETUPAPI.DLL` are not in `ALL_IMP_DLL` and they load just fine using this `LoadLibraryA` function. Why would it stop working on `SHELL32.DLL`?

In any case, I'll try again by adding those DLLs in ALL_IMP_DLL (if there's a better way than creating a new class that extends the Sandbox class, I'm all ears).

Thanks again!

**Comment by tomchop** [comment]
I ran some tests again, this time by putting all DLLs in `ALL_IMP_DLL` and I still get the same error. I've done a little more digging, and found some strange behaviour:

I made some modifications in `LoadLibraryA` to change `EIP` as I want to. When I change it to `0x40195F` (the `push` just before), the jitter loops indefinitely (it goes through the same instructions, then I change `EIP` again, etc.), which is the expected behaviour. When I manually set it to what the return address should be (`0x401968`), I get the error mentioned above.

If I manually set EIP to the instruction right after `0x401968` (in this case, `0x40196d`), the jitter continues for a while until hitting another similar error:

```
RAX 0000000077DD0000 RBX 0000000077F60000 RCX 0000000000000000 RDX 0000000000000000
RSI 000000007C801D7B RDI 0000000071126000 RSP 000000000132F5AC RBP 0000000077920000
zf 0000000000000000 nf 0000000000000000 of 0000000000000000 cf 0000000000000000
RIP 00000000004019A7
004019A9 JNZ        loc_00000000004019B2:0x004019b2`
WARNING: address 0x64000030 is not mapped in virtual memory:
WARNING: [Errno cannot get mem ad] 0x64000030L
WARNING: cannot disasm at 64000030
```
Any ideas?
